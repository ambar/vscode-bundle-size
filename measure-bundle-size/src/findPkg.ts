import escalade from 'escalade'
import {promises as fs} from 'fs'
import {createRequire} from 'module'
import path from 'path'

export type Pkg = {
  name: string
  version: string
  description?: string
  homepage?: string
  repository?: string
  dependencies?: Record<string, string>
  peerDependencies?: Record<string, string>
}

export const findPkg = (file: string) => {
  return escalade(file, (_, names) => {
    if (names.includes('package.json')) {
      return 'package.json'
    }
  })
}

const readJson = async (file: string) => {
  const content = await fs.readFile(file, 'utf-8')
  return JSON.parse(content) as Record<string, unknown>
}

const pkgName = 'package.json'
export const findPkgByName = (
  file: string,
  moduleName: string,
  callback?: (file: string, pkg: Pkg) => void
) => {
  return escalade(file, async (dir, names) => {
    if (names.includes(pkgName)) {
      const file = path.join(dir, pkgName)
      const json = (await readJson(file)) as Pkg
      callback?.(file, json)
      if (json.name === moduleName) {
        return pkgName
      }
    }
  })
}

export const findPkgs = async (modulePath: string, baseDir: string) => {
  const contextRequire = createRequire(path.resolve(baseDir, '<import>.js'))
  // ensure resolvable
  const moduleFile = contextRequire.resolve(modulePath)
  const isScoped = modulePath[0] === '@'
  const [p1, p2] = isScoped
    ? modulePath.split('/')
    : [modulePath.split('/')[0], '']
  const moduleName = p1 + (isScoped ? '/' : '') + p2
  let rootPkg: Pkg | void
  let rootPkgFile: string | void
  try {
    rootPkgFile = `${moduleName}/package.json`
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    rootPkg = require(rootPkgFile)
  } catch (_) {
    // This is probably: Package subpath './package.json' is not defined by "exports"
  }

  let modulePkg: Pkg | void
  let modulePkgFile: string | void
  if (moduleFile !== moduleName) {
    const pairs: [string, Pkg][] = []
    const pkgFile = await findPkgByName(moduleFile, moduleName, (x, y) => {
      pairs.push([x, y])
    })
    if (pkgFile && pairs.length > 0) {
      ;[rootPkgFile, rootPkg] = pairs[pairs.length - 1]
      if (pairs.length > 1) {
        ;[modulePkgFile, modulePkg] = pairs[0]
      }
    }
  }
  return {rootPkgFile, rootPkg, modulePkgFile, modulePkg}
}
