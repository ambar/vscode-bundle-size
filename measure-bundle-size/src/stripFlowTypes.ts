import {transform} from 'sucrase'

/**
 * Strip flow types from the given source code.
 *
 * Why sucrase is used instead of babel, bundle size of this package:
 * - `sucrase`: 848kb -> 1.1mb
 * - `@babel/core` (with `@babel/plugin-transform-flow-strip-types`): 848kb -> 5.7mb
 */
export function stripFlowTypes(source: string, filename?: string) {
  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call
  const result = transform(source, {
    transforms: ['flow', 'jsx', 'imports'],
    jsxRuntime: 'automatic',
    production: true,
    filePath: filename,
  })
  // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-member-access
  return result.code
}
