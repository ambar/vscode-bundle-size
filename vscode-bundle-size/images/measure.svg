<svg width="300" height="300" viewBox="0 0 300 300" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M248.909 81H50.0909C37.9409 81 28 89.85 28 100.667V179.333C28 190.15 37.9409 199 50.0909 199H248.909C261.059 199 271 190.15 271 179.333V100.667C271 89.85 261.059 81 248.909 81ZM248.909 179.333H50.0909V100.667H72.1818V140H94.2727V100.667H116.364V140H138.455V100.667H160.545V140H182.636V100.667H204.727V140H226.818V100.667H248.909V179.333Z" fill="#243733"/>
</svg>
