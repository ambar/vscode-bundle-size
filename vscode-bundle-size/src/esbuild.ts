import {exec} from 'child_process'
import {promisify} from 'util'

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type AnyFunction<T = void> = (...args: any[]) => T

function safeGet<T>(fn: AnyFunction<T>, defaultValue: T) {
  try {
    return fn()
  } catch (_) {
    return defaultValue
  }
}

const hasEsbuild = () => safeGet(() => !!require.resolve('esbuild'), false)
const command = promisify(exec)

/**
 * Get the right native executable of esbuild since it cannot be bundled
 * consider downloading: https://esbuild.github.io/getting-started/#download-a-build
 */
export const install = async (onInstall: AnyFunction) => {
  if (hasEsbuild()) {
    return
  }
  onInstall()
  let err: unknown
  for (const cmd of ['pnpm i --prod', 'yarn --prod', 'npm i --prod']) {
    await command(cmd, {cwd: __dirname}).catch((e) => {
      err = e
    })
    if (hasEsbuild()) {
      return
    }
  }
  return Promise.reject(err)
}
