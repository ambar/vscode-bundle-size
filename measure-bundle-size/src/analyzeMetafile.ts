import {sep} from 'path'
import bytes from 'bytes'
import * as esbuild from 'esbuild'

const groupBy = <T, K = string>(list: T[] = [], fn: (x: T) => K) => {
  const group = new Map<K, T[]>()
  for (const x of list) {
    const key = fn(x)
    const value = group.get(key)
    if (value) {
      value.push(x)
    } else {
      group.set(key, [x])
    }
  }
  return group
}

const splitModuleName = (path = '') => {
  const [p1, p2, ...rest] = path.split(sep)
  return p1.startsWith('@')
    ? [p1 + '/' + p2, rest.join(sep)]
    : [p1, [p2, ...rest].join(sep)]
}

const npmDir = 'node_modules'
const maxFilesInCell = 10

/**
 * Like `esbuild.analyzeMetafile`, but:
 * - render to markdown table
 * - more compact, accumulate files in same module
 * - strip common ancestor directory
 */
export default function analyzeMetafile(data: esbuild.Metafile) {
  const {outputs} = data
  const markdown: string[] = []

  for (const [outputName, outputStats] of Object.entries(outputs)) {
    const {entryPoint, inputs, bytes: totalBytes} = outputStats
    const list = Object.entries(inputs).map(([x, {bytesInOutput}]) => {
      if (x === entryPoint) {
        return {names: ['', x, ''], bytesInOutput}
      }
      const index = x.indexOf(npmDir)
      const parent = x.slice(0, index)
      const [moduleName, part2] = splitModuleName(
        x.slice(index + npmDir.length + 1)
      )
      return {names: [parent, moduleName, part2], bytesInOutput}
    })
    const group = groupBy(list, (x) => x.names[0])
    const nameGrouped = [...group].map(([_, entries]) => {
      return groupBy(entries, (x) => x.names[1])
      // TODO: show parent dir if needed, alert duplicates
      // return groupBy(entries, (x) => x.names[0] + x.names[1])
    })

    const head = ['Input', 'Files', 'Size', 'Percent']
    const lines = [
      `| ${head.join(' | ')} |`,
      `| ${head.map(() => '---').join(' | ')} |`,
    ]
    const leadingRow = {
      name: outputName,
      files: '',
      size: totalBytes,
      percent: '100%',
    }
    const rows = [leadingRow]
    for (const map of nameGrouped) {
      for (const [name, group] of map) {
        const files = group
          .map((x) => ({name: x.names[2], size: x.bytesInOutput}))
          .sort((a, b) => b.size - a.size)
        const bytesAcc = group.reduce((acc, x) => acc + x.bytesInOutput, 0)
        if (bytesAcc === 0) {
          continue
        }
        const percent = ((bytesAcc / totalBytes) * 100).toFixed(1) + '%'
        const getFiles = () => {
          if (files.length === 1) {
            return files[0].name
          }
          const format = (x: typeof files[0]) => `${x.name} (${bytes(x.size)})`
          if (files.length > maxFilesInCell) {
            const rest = files.slice(maxFilesInCell)
            const restSize = rest.reduce((acc, x) => acc + x.size, 0)
            return (
              files.slice(0, maxFilesInCell).map(format).join(', ') +
              `, and ${rest.length} other ${
                rest.length > 1 ? 'files' : 'file'
              } (${bytes(restSize)})`
            )
          }
          return files.map(format).join(', ')
        }
        rows.push({
          name: 'â†ª ' + name,
          files: getFiles(),
          size: bytesAcc,
          percent,
        })
      }
    }
    const maxLen = Math.max(...rows.map((r) => r.files.length))
    if (maxLen < 10) {
      // spread table
      leadingRow.files = '&nbsp; '.repeat(10)
    }
    const code = (text: string) => `\`${text}\``
    for (const row of rows.sort((a, b) => b.size - a.size)) {
      const cells = [code(row.name), row.files, bytes(row.size), row.percent]
      lines.push(`| ${cells.join(' | ')} |`)
    }
    markdown.push(lines.join('\n'))
  }

  return markdown.join('\n\n')
}
