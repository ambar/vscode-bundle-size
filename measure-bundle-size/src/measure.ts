import {promises as fs} from 'fs'
import path from 'path'
import bytes from 'bytes'
import Debug from 'debug'
import gzipSize from 'gzip-size'
import * as esbuild from 'esbuild'
import {builtinModules} from 'module'
import {parse, ImportInfo, exportImported} from './parse'
import analyzeMetafile from './analyzeMetafile'
import {findPkg, findPkgs, Pkg} from './findPkg'

const log = Debug('measure-bundle-size')

export const hasFile = async (file: string) =>
  Boolean(await fs.stat(file).catch(() => null))

export const pick = <T>(obj: T, keys: (keyof T)[]) =>
  keys.reduce((acc, k) => ((acc[k] = obj[k]), acc), {} as T)

const reBuiltin = RegExp(`^node:|^(${builtinModules.join('|')})(/|$)`)
/** ignore nested core modules  */
const builtinExternalPlugin: esbuild.Plugin = {
  name: 'external',
  setup(build) {
    build.onResolve({filter: reBuiltin, namespace: 'file'}, (args) => ({
      path: args.path,
      external: true,
    }))
  },
}
const reNonRelative = /^[a-z@]/
/** ignore nested peer deps  */
const peerExternalPlugin: esbuild.Plugin = {
  name: 'external',
  setup(build) {
    type CacheValue = Pkg | null
    const pkgCache = new Map<string, Promise<CacheValue>>()
    const externalsMap = new Map<string, Set<string>>()
    build.onResolve(
      {filter: reNonRelative, namespace: 'file'},
      async (args) => {
        if (reBuiltin.test(args.path)) {
          return null
        }
        // should passed by esbuild: https://github.com/evanw/esbuild/issues/978
        let pkg: CacheValue
        const cacheKey = args.resolveDir
        if (pkgCache.has(cacheKey)) {
          pkg = (await pkgCache.get(cacheKey)) as CacheValue
        } else {
          const result = findPkg(cacheKey).then((pkgFile) =>
            // eslint-disable-next-line @typescript-eslint/no-var-requires
            pkgFile ? (require(pkgFile) as Pkg) : null
          )
          pkgCache.set(cacheKey, result)
          pkg = await result
        }
        if (pkg?.peerDependencies && args.path in pkg.peerDependencies) {
          let externals = externalsMap.get(pkg.name)
          if (!externals) {
            externals = new Set()
            externalsMap.set(pkg.name, externals)
          }
          externals.add(args.path)
          return {
            path: args.path,
            external: true,
          }
        }
      }
    )
    build.onEnd(() => {
      if (externalsMap.size) {
        log(`peer externals`, externalsMap)
      }
    })
  },
}

type StatsOpt = boolean | 'tree' | 'table'

type BundleResult = {
  size: number
  zippedSize: number
  human: {size: string; zippedSize: string}
  pkg?: Pkg
  pkgFile?: string
  modulePkg?: Pkg
  modulePkgFile?: string
  stats?: string
}

const bundleCache = new Map<string, BundleResult>()
const genCacheKey = (pkg: Pkg, names: string) =>
  `${pkg.name}:${pkg.version}:${names}`

const timeMark = <T extends string>() => {
  const startTimes = new Map<T, number>()
  const values = new Map<T, number>()
  return {
    start(label: T) {
      startTimes.set(label, Date.now())
    },
    end(label: T) {
      const startTime = startTimes.get(label)
      if (!startTime) {
        throw new Error(`No label: ${label}`)
      }
      values.set(label, Date.now() - startTime)
    },
    print() {
      return [...values].map(([k, v]) => `${k} in ${v}ms`).join(', ')
    },
  }
}

const pickPkgFields = (pkg: Pkg) => {
  return pick<Pkg>(pkg, [
    'name',
    'version',
    'description',
    'homepage',
    'dependencies',
    'peerDependencies',
  ])
}

const bundle = async (
  statement: string,
  importInfo: ImportInfo,
  {
    baseDir,
    projectPkgFile,
    stats: statsOpt = false,
    cache: cacheOpt = false,
  }: {
    baseDir: string
    projectPkgFile?: string | void
    stats?: StatsOpt
    cache?: boolean
  }
): Promise<BundleResult> => {
  const bundleMark = timeMark<'bundle' | 'zip' | 'analyze'>()
  const modulePath = importInfo.from
  const [part1] = modulePath.split('/')
  if (!/^[a-z@]/.test(part1)) {
    // file path, alias
    throw new Error('Skip non-npm packages')
  }
  if (reBuiltin.test(part1)) {
    throw new Error('Skip builtin modules')
  }
  const entryInput = `${statement}\n${exportImported(importInfo)}`
  type T = Awaited<ReturnType<typeof findPkgs>>
  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
  const {rootPkgFile, rootPkg, modulePkgFile, modulePkg} = await findPkgs(
    modulePath,
    baseDir
  ).catch(() => {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-return
    return {} as T
  })

  const cacheKey =
    cacheOpt && rootPkg !== undefined && genCacheKey(rootPkg, entryInput)
  if (cacheOpt && cacheKey) {
    const cache = bundleCache.get(cacheKey)
    if (cache) {
      log('using cache:', importInfo.from)
      return cache
    }
  }

  bundleMark.start('bundle')
  // TODO: alert indirect deps, eg.: react-redux > react-dom
  // exclude peer deps, eg: react-router-dom > react
  const external = rootPkg ? Object.keys({...rootPkg.peerDependencies}) : []
  // reduce depth of relative path in `metafile`, can't resolve directly to `node_modules` because it may not exist (or in monorepo)
  const workingDir = projectPkgFile ? path.dirname(projectPkgFile) : baseDir
  /**
   * - build in memory, no file writing
   * - respect 'browser' field in package.json
   * - production build, respect `NODE_ENV`
   *
   * TODO: add loader for `.node`
   * @see https://esbuild.github.io/api/#build-api
   */
  const buildResult = await esbuild.build({
    stdin: {
      loader: 'ts',
      contents: entryInput,
      resolveDir: workingDir,
      sourcefile: '<import>',
    },
    outfile: '<bundle>.js',
    absWorkingDir: workingDir,
    plugins: [builtinExternalPlugin, peerExternalPlugin],
    external,
    target: 'esnext',
    format: 'esm',
    platform: 'browser',
    metafile: statsOpt !== false,
    minify: true,
    bundle: true,
    write: false,
  })
  bundleMark.end('bundle')

  // analyze on hover?
  let stats
  if (buildResult.metafile) {
    bundleMark.start('analyze')
    if (statsOpt === 'tree') {
      stats = await esbuild.analyzeMetafile(buildResult.metafile)
    } else if (statsOpt === 'table') {
      stats = analyzeMetafile(buildResult.metafile)
    }
    bundleMark.end('analyze')
    // if (stats) {
    //   log('stats:\n' + stats)
    // }
  }
  // concat <bundle>.js, <bundle>.css, etc
  bundleMark.start('zip')
  const sizes = await Promise.all(
    buildResult.outputFiles.map(async (file) => {
      const itsSize = file.contents.length
      return [
        itsSize,
        itsSize === 0 ? 0 : await gzipSize(Buffer.from(file.contents)),
      ] as [number, number]
    })
  )
  bundleMark.end('zip')
  let size = 0
  let zippedSize = 0
  for (const [s1, s2] of sizes) {
    size += s1
    zippedSize += s2
  }
  const human = {size: bytes(size), zippedSize: bytes(zippedSize)}
  log(`${bundleMark.print()}, ${statement}, ${human.size}`)
  const toRelative = path.relative.bind(null, process.cwd())
  const result = {
    size,
    zippedSize,
    human,
    pkg: rootPkg ? pickPkgFields(rootPkg) : undefined,
    pkgFile: rootPkgFile ? toRelative(rootPkgFile) : undefined,
    modulePkg: modulePkg ? pickPkgFields(modulePkg) : undefined,
    modulePkgFile: modulePkgFile ? toRelative(modulePkgFile) : undefined,
    stats,
  }
  if (cacheOpt && cacheKey) {
    bundleCache.set(cacheKey, result)
  }
  return result
}

export type MeasureResult = {
  importInfo: ImportInfo
  result?: BundleResult
  error?: Error | unknown
}

type MeasureOptions = {
  debug?: boolean
  stats?: StatsOpt
  cache?: boolean
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  log?: (...args: any[]) => void
  workspaceFolder?: string
}

export const measure = async (
  input: string,
  fileName?: string | null,
  opts: MeasureOptions = {}
): Promise<MeasureResult[]> => {
  if (opts.debug) {
    log.enabled = true
  }
  if (opts.log) {
    log.log = opts.log
  }

  log('process', fileName)
  const baseDir =
    fileName && (await hasFile(fileName))
      ? path.dirname(fileName)
      : opts.workspaceFolder && (await hasFile(opts.workspaceFolder))
      ? opts.workspaceFolder
      : null
  if (!baseDir) {
    throw new Error('Can not resolve `fileName` or `workspaceFolder`')
  }

  const projectPkgFile = await findPkg(baseDir)
  const parseMark = timeMark<'parse'>()
  parseMark.start('parse')
  const result = parse(input)
  parseMark.end('parse')
  log(parseMark.print())
  const bundleOpts = {
    baseDir,
    projectPkgFile,
    stats: opts.stats,
    cache: opts.cache,
  }
  return Promise.all(
    result.imports.map(async (importInfo) => {
      const statement = input.substring(importInfo.start, importInfo.end)
      return bundle(statement, importInfo, bundleOpts).then(
        (result) => ({
          importInfo,
          result,
          error: void 0,
        }),
        (error: Error) => ({
          importInfo,
          result: void 0,
          error,
        })
      )
    })
  )
}
