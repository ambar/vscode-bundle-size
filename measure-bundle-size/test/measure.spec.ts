/* eslint-disable @typescript-eslint/no-non-null-assertion */
import path from 'path'
import dedent from 'dedent'
import {measure} from '../src/measure'

const inputs = [
  // skip node core
  `
    import path from 'path'
    import fs from 'fs/promise'
    import url from 'node:url'
    `,

  // skip relative
  `
    import * as utils from './utils'
    `,
  // CJS package: path import
  `
    import map from 'lodash/map'
    `,
  // CJS package: main entry
  `
    import {map} from 'lodash'
    `,
  // ESM package: main entry
  `
    import {lightFormat} from 'date-fns'
    `,
  // namespace: usingProps=0
  `
    import * as dateFns from 'date-fns'
    `,
  // namespace: usingProps=0, has scope name
  `
    import * as dateFns from 'date-fns'
    {
      let dateFns = {}
      {
        dateFns.addDays
      }
      void (function(dateFns) {
        dateFns.addDays()
      }())
    }
    `,
  // namespace: usingProps=1
  `
    import * as dateFns from 'date-fns'
    dateFns.addDays
    `,
  // tree-shaking NODE_ENV=production
  `
    import React from 'react'
  `,
  // exclude peer deps (react)
  `
    import 'react-inline-center'
  `,
  // scoped path
  `
    import factory from '@ambarli/alias/factory'
  `,
  // bundle CSS
  `
    import 'sanitize.css'
    import 'sanitize.css/forms.css'
  `,
]

test.each(inputs)('measure %p', async (input) => {
  expect(await measure(dedent(input), __filename)).toMatchSnapshot()
})

test('throw', async () => {
  await expect(measure(``)).rejects.toThrow(
    /Can not resolve `fileName` or `workspaceFolder`/
  )
})

test('debug', async () => {
  const log = jest.fn()
  expect(await measure(``, __filename, {debug: true, log})).toEqual([])
  expect(log).toBeCalled()
})

test('analyze', async () => {
  // resolved to project root
  const [r] = await measure(`import React from 'react'`, __filename, {
    stats: 'table',
  })
  expect(r.result!.stats).toMatchSnapshot()
})

test('alias in tsconfig', async () => {
  const [r] = await measure(`import magic from 'pkg-mylib'`, null, {
    workspaceFolder: path.resolve(__dirname, 'fixtures/alias-mylib'),
    stats: 'table',
  })
  expect(r.result!.stats).toMatchSnapshot()
})

test('add ellipsis for files', async () => {
  const [r] = await measure(`import debounce from 'lodash/debounce'`, null, {
    workspaceFolder: __dirname,
    stats: 'table',
  })
  expect(r.result!.stats).toMatchSnapshot()
})

test('no package.json in exports', async () => {
  const [r] = await measure(`import escalade from 'escalade'`, null, {
    workspaceFolder: __dirname,
    stats: 'table',
  })
  expect(r.result!.stats).toMatchSnapshot()
})

test('parse flow type', async () => {
  const [r] = await measure(
    `import {registerAsset} from '@react-native/assets/registry'`,
    null,
    {
      workspaceFolder: __dirname,
      stats: 'table',
    }
  )
  expect(r.result!.stats).toMatchSnapshot()
})
