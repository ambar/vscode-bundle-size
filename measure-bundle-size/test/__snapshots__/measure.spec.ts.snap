// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add ellipsis for files 1`] = `
"| Input | Files | Size | Percent |
| --- | --- | --- | --- |
| \`<bundle>.js\` |  | 3.41KB | 100% |
| \`↪ lodash\` | debounce.js (998B), toNumber.js (399B), _getRawTag.js (245B), _baseGetTag.js (201B), _root.js (139B), isSymbol.js (130B), _trimmedEndIndex.js (115B), _baseTrim.js (112B), isObject.js (107B), _objectToString.js (104B), and 4 other files (320B) | 2.8KB | 82.3% |
| \`↪ <import>\` |  | 16B | 0.5% |"
`;

exports[`alias in tsconfig 1`] = `
"| Input | Files | Size | Percent |
| --- | --- | --- | --- |
| \`<bundle>.js\` | &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  | 29B | 100% |
| \`↪ pkg-mylib\` | main.js | 9B | 31.0% |"
`;

exports[`analyze 1`] = `
"| Input | Files | Size | Percent |
| --- | --- | --- | --- |
| \`<bundle>.js\` |  | 7.67KB | 100% |
| \`↪ react\` | cjs/react.production.min.js (5.82KB), index.js (46B) | 5.87KB | 76.5% |
| \`↪ object-assign\` | index.js | 956B | 12.2% |
| \`↪ <import>\` |  | 14B | 0.2% |"
`;

exports[`measure "
    import 'react-inline-center'
  " 1`] = `
Array [
  Object {
    "error": undefined,
    "importInfo": Object {
      "end": 28,
      "from": "react-inline-center",
      "start": 0,
    },
    "result": Object {
      "human": Object {
        "size": "22B",
        "zippedSize": "42B",
      },
      "modulePkg": undefined,
      "modulePkgFile": undefined,
      "pkg": Object {
        "dependencies": undefined,
        "description": "Centering inline element",
        "homepage": undefined,
        "name": "react-inline-center",
        "peerDependencies": Object {
          "react": "*",
        },
        "version": "1.0.1",
      },
      "pkgFile": "node_modules/react-inline-center/package.json",
      "size": 22,
      "stats": undefined,
      "zippedSize": 42,
    },
  },
]
`;

exports[`measure "
    import 'sanitize.css'
    import 'sanitize.css/forms.css'
  " 1`] = `
Array [
  Object {
    "error": undefined,
    "importInfo": Object {
      "end": 21,
      "from": "sanitize.css",
      "start": 0,
    },
    "result": Object {
      "human": Object {
        "size": "2.15KB",
        "zippedSize": "936B",
      },
      "modulePkg": undefined,
      "modulePkgFile": undefined,
      "pkg": Object {
        "dependencies": undefined,
        "description": "A best-practices CSS foundation",
        "homepage": "https://github.com/csstools/sanitize.css#readme",
        "name": "sanitize.css",
        "peerDependencies": undefined,
        "version": "13.0.0",
      },
      "pkgFile": "node_modules/sanitize.css/package.json",
      "size": 2203,
      "stats": undefined,
      "zippedSize": 936,
    },
  },
  Object {
    "error": undefined,
    "importInfo": Object {
      "end": 53,
      "from": "sanitize.css/forms.css",
      "start": 22,
    },
    "result": Object {
      "human": Object {
        "size": "529B",
        "zippedSize": "357B",
      },
      "modulePkg": undefined,
      "modulePkgFile": undefined,
      "pkg": Object {
        "dependencies": undefined,
        "description": "A best-practices CSS foundation",
        "homepage": "https://github.com/csstools/sanitize.css#readme",
        "name": "sanitize.css",
        "peerDependencies": undefined,
        "version": "13.0.0",
      },
      "pkgFile": "node_modules/sanitize.css/package.json",
      "size": 529,
      "stats": undefined,
      "zippedSize": 357,
    },
  },
]
`;

exports[`measure "
    import * as dateFns from 'date-fns'
    " 1`] = `
Array [
  Object {
    "error": undefined,
    "importInfo": Object {
      "end": 35,
      "from": "date-fns",
      "namespace": Object {
        "name": "dateFns",
        "usingProps": Array [],
      },
      "start": 0,
    },
    "result": Object {
      "human": Object {
        "size": "0B",
        "zippedSize": "0B",
      },
      "modulePkg": undefined,
      "modulePkgFile": undefined,
      "pkg": Object {
        "dependencies": Object {},
        "description": "Modern JavaScript date utility library",
        "homepage": undefined,
        "name": "date-fns",
        "peerDependencies": undefined,
        "version": "2.17.0",
      },
      "pkgFile": "node_modules/date-fns/package.json",
      "size": 0,
      "stats": undefined,
      "zippedSize": 0,
    },
  },
]
`;

exports[`measure "
    import * as dateFns from 'date-fns'
    {
      let dateFns = {}
      {
        dateFns.addDays
      }
      void (function(dateFns) {
        dateFns.addDays()
      }())
    }
    " 1`] = `
Array [
  Object {
    "error": undefined,
    "importInfo": Object {
      "end": 35,
      "from": "date-fns",
      "namespace": Object {
        "name": "dateFns",
        "usingProps": Array [],
      },
      "start": 0,
    },
    "result": Object {
      "human": Object {
        "size": "0B",
        "zippedSize": "0B",
      },
      "modulePkg": undefined,
      "modulePkgFile": undefined,
      "pkg": Object {
        "dependencies": Object {},
        "description": "Modern JavaScript date utility library",
        "homepage": undefined,
        "name": "date-fns",
        "peerDependencies": undefined,
        "version": "2.17.0",
      },
      "pkgFile": "node_modules/date-fns/package.json",
      "size": 0,
      "stats": undefined,
      "zippedSize": 0,
    },
  },
]
`;

exports[`measure "
    import * as dateFns from 'date-fns'
    dateFns.addDays
    " 1`] = `
Array [
  Object {
    "error": undefined,
    "importInfo": Object {
      "end": 35,
      "from": "date-fns",
      "namespace": Object {
        "name": "dateFns",
        "usingProps": Array [
          Object {
            "end": 51,
            "name": "addDays",
            "start": 36,
          },
        ],
      },
      "start": 0,
    },
    "result": Object {
      "human": Object {
        "size": "869B",
        "zippedSize": "506B",
      },
      "modulePkg": undefined,
      "modulePkgFile": undefined,
      "pkg": Object {
        "dependencies": Object {},
        "description": "Modern JavaScript date utility library",
        "homepage": undefined,
        "name": "date-fns",
        "peerDependencies": undefined,
        "version": "2.17.0",
      },
      "pkgFile": "node_modules/date-fns/package.json",
      "size": 869,
      "stats": undefined,
      "zippedSize": 506,
    },
  },
]
`;

exports[`measure "
    import * as utils from './utils'
    " 1`] = `
Array [
  Object {
    "error": [Error: Skip non-npm packages],
    "importInfo": Object {
      "end": 32,
      "from": "./utils",
      "namespace": Object {
        "name": "utils",
        "usingProps": Array [],
      },
      "start": 0,
    },
    "result": undefined,
  },
]
`;

exports[`measure "
    import {lightFormat} from 'date-fns'
    " 1`] = `
Array [
  Object {
    "error": undefined,
    "importInfo": Object {
      "end": 36,
      "from": "date-fns",
      "names": Object {
        "lightFormat": "lightFormat",
      },
      "start": 0,
    },
    "result": Object {
      "human": Object {
        "size": "2.43KB",
        "zippedSize": "1.13KB",
      },
      "modulePkg": undefined,
      "modulePkgFile": undefined,
      "pkg": Object {
        "dependencies": Object {},
        "description": "Modern JavaScript date utility library",
        "homepage": undefined,
        "name": "date-fns",
        "peerDependencies": undefined,
        "version": "2.17.0",
      },
      "pkgFile": "node_modules/date-fns/package.json",
      "size": 2489,
      "stats": undefined,
      "zippedSize": 1162,
    },
  },
]
`;

exports[`measure "
    import {map} from 'lodash'
    " 1`] = `
Array [
  Object {
    "error": undefined,
    "importInfo": Object {
      "end": 26,
      "from": "lodash",
      "names": Object {
        "map": "map",
      },
      "start": 0,
    },
    "result": Object {
      "human": Object {
        "size": "71.71KB",
        "zippedSize": "25.91KB",
      },
      "modulePkg": undefined,
      "modulePkgFile": undefined,
      "pkg": Object {
        "dependencies": undefined,
        "description": "Lodash modular utilities.",
        "homepage": "https://lodash.com/",
        "name": "lodash",
        "peerDependencies": undefined,
        "version": "4.17.21",
      },
      "pkgFile": "node_modules/lodash/package.json",
      "size": 73432,
      "stats": undefined,
      "zippedSize": 26531,
    },
  },
]
`;

exports[`measure "
    import React from 'react'
  " 1`] = `
Array [
  Object {
    "error": undefined,
    "importInfo": Object {
      "end": 25,
      "from": "react",
      "names": Object {
        "default": "React",
      },
      "start": 0,
    },
    "result": Object {
      "human": Object {
        "size": "7.67KB",
        "zippedSize": "3.21KB",
      },
      "modulePkg": undefined,
      "modulePkgFile": undefined,
      "pkg": Object {
        "dependencies": Object {
          "loose-envify": "^1.1.0",
          "object-assign": "^4.1.1",
        },
        "description": "React is a JavaScript library for building user interfaces.",
        "homepage": "https://reactjs.org/",
        "name": "react",
        "peerDependencies": undefined,
        "version": "17.0.2",
      },
      "pkgFile": "node_modules/react/package.json",
      "size": 7851,
      "stats": undefined,
      "zippedSize": 3288,
    },
  },
]
`;

exports[`measure "
    import factory from '@ambarli/alias/factory'
  " 1`] = `
Array [
  Object {
    "error": undefined,
    "importInfo": Object {
      "end": 44,
      "from": "@ambarli/alias/factory",
      "names": Object {
        "default": "factory",
      },
      "start": 0,
    },
    "result": Object {
      "human": Object {
        "size": "637B",
        "zippedSize": "372B",
      },
      "modulePkg": undefined,
      "modulePkgFile": undefined,
      "pkg": Object {
        "dependencies": undefined,
        "description": "Get aliases",
        "homepage": undefined,
        "name": "@ambarli/alias",
        "peerDependencies": undefined,
        "version": "0.0.0",
      },
      "pkgFile": "node_modules/@ambarli/alias/package.json",
      "size": 637,
      "stats": undefined,
      "zippedSize": 372,
    },
  },
]
`;

exports[`measure "
    import map from 'lodash/map'
    " 1`] = `
Array [
  Object {
    "error": undefined,
    "importInfo": Object {
      "end": 28,
      "from": "lodash/map",
      "names": Object {
        "default": "map",
      },
      "start": 0,
    },
    "result": Object {
      "human": Object {
        "size": "21.1KB",
        "zippedSize": "8.29KB",
      },
      "modulePkg": undefined,
      "modulePkgFile": undefined,
      "pkg": Object {
        "dependencies": undefined,
        "description": "Lodash modular utilities.",
        "homepage": "https://lodash.com/",
        "name": "lodash",
        "peerDependencies": undefined,
        "version": "4.17.21",
      },
      "pkgFile": "node_modules/lodash/package.json",
      "size": 21611,
      "stats": undefined,
      "zippedSize": 8489,
    },
  },
]
`;

exports[`measure "
    import path from 'path'
    import fs from 'fs/promise'
    import url from 'node:url'
    " 1`] = `
Array [
  Object {
    "error": [Error: Skip builtin modules],
    "importInfo": Object {
      "end": 23,
      "from": "path",
      "names": Object {
        "default": "path",
      },
      "start": 0,
    },
    "result": undefined,
  },
  Object {
    "error": [Error: Skip builtin modules],
    "importInfo": Object {
      "end": 51,
      "from": "fs/promise",
      "names": Object {
        "default": "fs",
      },
      "start": 24,
    },
    "result": undefined,
  },
  Object {
    "error": [Error: Skip builtin modules],
    "importInfo": Object {
      "end": 78,
      "from": "node:url",
      "names": Object {
        "default": "url",
      },
      "start": 52,
    },
    "result": undefined,
  },
]
`;

exports[`no package.json in exports 1`] = `
"| Input | Files | Size | Percent |
| --- | --- | --- | --- |
| \`<bundle>.js\` |  | 308B | 100% |
| \`↪ escalade\` | dist/index.mjs | 285B | 92.5% |"
`;
